pkg/log/zap/logger_test.go

```diff
@@ -0,0 +1,171 @@
+				set: false,
+			},
+			inLevel: levelValue{
+				set: false,
+			},
+			inSample: sampleValue{
+				set: false,
+			},
+			expected: config{
+				encoder: consoleEncoder(),
+				level:   zap.NewAtomicLevelAt(zap.DebugLevel),
+				opts:    append(opts, zap.Development(), zap.AddStacktrace(zap.ErrorLevel)),
+				sample:  false,
+			},
+		},
+		{
+			name:    "development off",
+			inDevel: false,
+			inEncoder: encoderValue{
+				set: false,
+			},
+			inLevel: levelValue{
+				set: false,
+			},
+			inSample: sampleValue{
+				set: false,
+			},
+			expected: config{
+				encoder: jsonEncoder(),
+				level:   zap.NewAtomicLevelAt(zap.InfoLevel),
+				opts:    append(opts, zap.AddStacktrace(zap.WarnLevel)),
+				sample:  true,
+			},
+		},
+		{
+			name:    "set encoder",
+			inDevel: false,
+			inEncoder: encoderValue{
+				set:     true,
+				encoder: consoleEncoder(),
+			},
+			inLevel: levelValue{
+				set: false,
+			},
+			inSample: sampleValue{
+				set: false,
+			},
+			expected: config{
+				encoder: jsonEncoder(),
+				level:   zap.NewAtomicLevelAt(zap.InfoLevel),
+				opts:    append(opts, zap.AddStacktrace(zap.WarnLevel)),
+				sample:  true,
+			},
+		},
+		{
+			name:    "set level using level constant",
+			inDevel: false,
+			inEncoder: encoderValue{
+				set: false,
+			},
+			inLevel: levelValue{
+				set:   true,
+				level: zapcore.ErrorLevel,
+			},
+			inSample: sampleValue{
+				set: false,
+			},
+			expected: config{
+				encoder: jsonEncoder(),
+				level:   zap.NewAtomicLevelAt(zap.ErrorLevel),
+				opts:    append(opts, zap.AddStacktrace(zap.WarnLevel)),
+				sample:  true,
+			},
+		},
+		{
+			name:    "set level using custom level",
+			inDevel: false,
+			inEncoder: encoderValue{
+				set: false,
+			},
+			inLevel: levelValue{
+				set:   true,
+				level: zapcore.Level(-10),
+			},
+			inSample: sampleValue{
+				set: false,
+			},
+			expected: config{
+				encoder: jsonEncoder(),
+				level:   zap.NewAtomicLevelAt(zapcore.Level(-10)),
+				opts:    append(opts, zap.AddStacktrace(zap.WarnLevel)),
+				sample:  false,
+			},
+		},
+		{
+			name:    "set level using custom level, sample override not possible",
+			inDevel: false,
+			inEncoder: encoderValue{
+				set: false,
+			},
+			inLevel: levelValue{
+				set:   true,
+				level: zapcore.Level(-10),
+			},
+			inSample: sampleValue{
+				set:    true,
+				sample: true,
+			},
+			expected: config{
+				encoder: jsonEncoder(),
+				level:   zap.NewAtomicLevelAt(zapcore.Level(-10)),
+				opts:    append(opts, zap.AddStacktrace(zap.WarnLevel)),
+				sample:  false,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			development = tc.inDevel
+			encoderVal = tc.inEncoder
+			levelVal = tc.inLevel
+			sampleVal = tc.inSample
+
+			cfg := getConfig()
+			assert.Equal(t, tc.expected.level, cfg.level)
+			assert.Equal(t, len(tc.expected.opts), len(cfg.opts))
+			assert.Equal(t, tc.expected.sample, cfg.sample)
```

Comment:
Should we also actually build the logger from the config and attempt to call `log.V(10).Info(...)` to make sure we don't panic when debug logging is enabled?

Reply:
That seems like an e2e test. the unit test is testing the code is behaving how we expected. 

I could see us enabling this in the ansible e2e test to help debug CI failures. I would imagine this would be a follow-up PR. 

thoughts?

Reply:
It seems like something we could do pretty easily in a `func TestLoggerDoesntPanic(t *testing.T) {}`-style test, and I think such a test would fall under the umbrella of "code is behaving how we expected".

But I'm on board with doing something like this as a follow up.

Reply:
@joelanford @shawn-hurley I can add a more thorough test in a follow up if that is acceptable.